go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,562 @@
+		nodeName:        nodeName,
+		manageDBServers: manageDBServers,
+		isLeader:        false,
+		leaderElector:   nil,
+	}
+}
+
+// StartHAMasterCluster runs the replication controller
+func (hacluster *OvnHAMasterController) StartHAMasterCluster() error {
+	if hacluster.manageDBServers {
+		// Always demote the OVN DBs to backmode mode.
+		// After the leader election, the leader will promote the OVN Dbs
+		// to become active.
+		err := hacluster.DemoteOVNDbs(invalidIPAddress, "6641", "6642")
+		if err != nil {
+			// If we are not able to communicate to the OVN ovsdb-servers,
+			// then it is better to return than continue.
+			// cmd/ovnkube.go will panic if this function returns error.
+			return err
+		}
+	}
+
+	// Set up leader election process first
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		haLeaderLockName,
+		hacluster.kubeClient.CoreV1(),
+		nil,
+		resourcelock.ResourceLockConfig{
+			Identity:      hacluster.nodeName,
+			EventRecorder: nil,
+		})
+
+	if err != nil {
+		return err
+	}
+
+	HAClusterNewLeader := func(nodeName string) {
+		logrus.Infof(nodeName + " is the new leader")
+		wasLeader := hacluster.isLeader
+
+		if hacluster.nodeName == nodeName {
+			logrus.Infof(" I (" + hacluster.nodeName + ") won the election. In active mode")
+			err = hacluster.ConfigureAsActive(nodeName)
+			if err != nil {
+				logrus.Errorf(err.Error())
+				if hacluster.manageDBServers {
+					// Stop ovn-northd before panicing.
+					_, _, _ = util.RunOVNNorthAppCtl("exit")
+				}
+				panic(err.Error())
+			}
+			hacluster.isLeader = true
+		} else {
+			logrus.Infof(" I (" + hacluster.nodeName + ") lost the election. In Standby mode")
+			if wasLeader {
+				//This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				if hacluster.manageDBServers {
+					// Stop ovn-northd and then exit.
+					_, _, _ = util.RunOVNNorthAppCtl("exit")
+				}
+				logrus.Infof("I (" + hacluster.nodeName + ") am no longer a leader. Exiting")
+				os.Exit(1)
+			}
+			ep, er := hacluster.cluster.Kube.GetEndpoint(config.Kubernetes.OVNConfigNamespace, ovnkubeDbEp)
+			if er == nil {
+				er = hacluster.ConfigureAsStandby(ep)
+				if er != nil {
+					logrus.Errorf(er.Error())
+					if hacluster.manageDBServers {
+						// Stop ovn-northd and then exit
+						_, _, _ = util.RunOVNNorthAppCtl("exit")
+					}
+					panic(er.Error())
+				}
+			}
+		}
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: leaseDuration,
+		RenewDeadline: renewDeadline,
+		RetryPeriod:   retryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {},
+			OnStoppedLeading: func() {
```

Comment:
Since this code is the same as what's in HAClusterNewLeader() for this case, could we just make a small function with the code and call that function from both places?

Reply:
Sounds good. Done